src/Utilities/DoNotCatchGeneralUnlessRethrown.cs

```diff
@@ -0,0 +1,149 @@
+        }
+
+        private static ICollection<INamedTypeSymbol> GetDisallowedCatchTypes(Compilation compilation)
+        {
+            var disallowed = new List<INamedTypeSymbol>();
+            disallowed.Add(WellKnownTypes.Object(compilation));
+            disallowed.Add(WellKnownTypes.Exception(compilation));
+            disallowed.Add(WellKnownTypes.SystemException(compilation));
+            return disallowed;
+        }
+
+        /// <summary>
+        /// Walks an IOperation tree to find catch blocks that handle general types without rethrowing them.
+        /// </summary>
+        private class DisallowGeneralCatchUnlessRethrowWalker : OperationWalker
+        {
+            private readonly ICollection<INamedTypeSymbol> _disallowedCatchTypes;
+            private readonly bool _checkAnonymousFunctions;
+            private readonly Stack<bool> _seenRethrowInCatchClauses = new Stack<bool>();
+
+            public ISet<ICatchClauseOperation> CatchClausesForDisallowedTypesWithoutRethrow { get; } = new HashSet<ICatchClauseOperation>();
+
+            public DisallowGeneralCatchUnlessRethrowWalker(ICollection<INamedTypeSymbol> disallowedCatchTypes, bool checkAnonymousFunctions)
+            {
+                _disallowedCatchTypes = disallowedCatchTypes;
+                _checkAnonymousFunctions = checkAnonymousFunctions;
+            }
+
+            public override void VisitAnonymousFunction(IAnonymousFunctionOperation operation)
+            {
+                if (_checkAnonymousFunctions)
+                {
+                    base.VisitAnonymousFunction(operation);
+                }
+            }
+
+            public override void VisitCatchClause(ICatchClauseOperation operation)
+            {
+                // No diagnostic if an exception filter is specified
+                if (operation.Filter == null)
+                {
+                    _seenRethrowInCatchClauses.Push(false);
+
+                    Visit(operation.Filter);
+                    Visit(operation.Handler);
+
+                    bool seenRethrow = _seenRethrowInCatchClauses.Pop();
+
+                    if (IsCaughtTypeDisallowed(operation.ExceptionType) && !seenRethrow)
+                    {
+                        CatchClausesForDisallowedTypesWithoutRethrow.Add(operation);
+                    }
+                }
+            }
+
+            public override void VisitThrow(IThrowOperation operation)
+            {
+                if (operation.Exception == null && _seenRethrowInCatchClauses.Count > 0 && !_seenRethrowInCatchClauses.Peek())
+                {
+                    _seenRethrowInCatchClauses.Pop();
+                    _seenRethrowInCatchClauses.Push(true);
+                }
+
+                base.VisitThrow(operation);
+            }
+
+            private bool IsCaughtTypeDisallowed(ITypeSymbol caughtType)
+            {
+                return caughtType == null || _disallowedCatchTypes.Any(type => caughtType == type);
```

Comment:
>== [](start = 90, length = 2)

We should use `.Equals` invocation to compare type symbols.

Reply:
Done